File: app/src/main/java/org/dhis2/usescases/teiDashboard/TeiDashboardModule.java
package org.dhis2.usescases.teiDashboard;

import androidx.annotation.NonNull;

import org.dhis2.commons.di.dagger.PerActivity;
import org.dhis2.commons.prefs.PreferenceProvider;
import org.dhis2.data.forms.EnrollmentFormRepository;
import org.dhis2.data.forms.FormRepository;
import org.dhis2.form.data.RulesRepository;
import org.dhis2.data.forms.dataentry.EnrollmentRuleEngineRepository;
import org.dhis2.data.forms.dataentry.RuleEngineRepository;
import org.dhis2.commons.schedulers.SchedulerProvider;
import org.dhis2.utils.analytics.AnalyticsHelper;
import org.dhis2.commons.matomo.MatomoAnalyticsController;
import org.dhis2.utils.customviews.navigationbar.NavigationPageConfigurator;
import org.dhis2.commons.filters.FilterManager;
import org.dhis2.commons.resources.ResourceManager;
import org.hisp.dhis.android.core.D2;

import dagger.Module;
import dagger.Provides;
import dhis2.org.analytics.charts.Charts;

/**
 * QUADRAM. Created by ppajuelo on 30/11/2017.
 */
@Module
public class TeiDashboardModule {

    public final String programUid;
    public final String teiUid;
    private final TeiDashboardContracts.View view;
    private final String enrollmentUid;
    private final boolean isPortrait;

    public TeiDashboardModule(TeiDashboardContracts.View view, String teiUid, String programUid, String enrollmentUid, boolean isPortrait) {
        this.view = view;
        this.teiUid = teiUid;
        this.programUid = programUid;
        this.enrollmentUid = enrollmentUid;
        this.isPortrait = isPortrait;
    }

    @Provides
    @PerActivity
    TeiDashboardContracts.View provideView(TeiDashboardMobileActivity mobileActivity) {
        return mobileActivity;
    }

    @Provides
    @PerActivity
    TeiDashboardContracts.Presenter providePresenter(DashboardRepository dashboardRepository,
                                                     SchedulerProvider schedulerProvider,
                                                     AnalyticsHelper analyticsHelper,
                                                     PreferenceProvider preferenceProvider,
                                                     FilterManager filterManager,
                                                     MatomoAnalyticsController matomoAnalyticsController) {
        return new TeiDashboardPresenter(view,
                teiUid,
                programUid,
                enrollmentUid,
                dashboardRepository,
                schedulerProvider,
                analyticsHelper,
                preferenceProvider,
                filterManager,
                matomoAnalyticsController);
    }

    @Provides
    @PerActivity
    DashboardRepository dashboardRepository(D2 d2, Charts charts, ResourceManager resources, TeiAttributesProvider teiAttributesProvider) {
        return new DashboardRepositoryImpl(d2, charts, teiUid, programUid, enrollmentUid, resources, teiAttributesProvider);
    }

    @Provides
    @PerActivity
    RulesRepository rulesRepository(@NonNull D2 d2) {
        return new RulesRepository(d2);
    }

    @Provides
    @PerActivity
    FormRepository formRepository(
            @NonNull RulesRepository rulesRepository,
            D2 d2) {
        String enrollmentUidToUse = enrollmentUid != null ? enrollmentUid : "";
        return new EnrollmentFormRepository(rulesRepository, enrollmentUidToUse, d2);
    }

    @Provides
    @PerActivity
    RuleEngineRepository ruleEngineRepository(D2 d2, FormRepository formRepository) {
        String enrollmentUidToUse = enrollmentUid != null ? enrollmentUid : "";
        return new EnrollmentRuleEngineRepository(formRepository, enrollmentUidToUse, d2);
    }

    @Provides
    @PerActivity
    NavigationPageConfigurator pageConfigurator(DashboardRepository dashboardRepository) {
        return new TeiDashboardPageConfigurator(dashboardRepository, isPortrait);
    }

    @Provides
    @PerActivity
    TeiAttributesProvider teiAttributesProvider(D2 d2) {
        return new TeiAttributesProvider(d2);
    }
}

File: app/src/main/java/org/dhis2/usescases/teiDashboard/TeiDashboardModule.kt
package org.dhis2.usescases.teiDashboard

import dagger.Module
import dagger.Provides
import dhis2.org.analytics.charts.Charts
import org.dhis2.commons.data.ProgramConfigurationRepository
import org.dhis2.commons.di.dagger.PerActivity
import org.dhis2.commons.featureconfig.data.FeatureConfigRepository
import org.dhis2.commons.matomo.MatomoAnalyticsController
import org.dhis2.commons.prefs.PreferenceProvider
import org.dhis2.commons.resources.EventResourcesProvider
import org.dhis2.commons.resources.MetadataIconProvider
import org.dhis2.commons.resources.ResourceManager
import org.dhis2.commons.schedulers.SchedulerProvider
import org.dhis2.commons.viewmodel.DispatcherProvider
import org.dhis2.form.data.metadata.EnrollmentConfiguration
import org.dhis2.mobileProgramRules.EvaluationType
import org.dhis2.mobileProgramRules.RuleEngineHelper
import org.dhis2.usescases.enrollment.DateEditionWarningHandler
import org.dhis2.utils.analytics.AnalyticsHelper
import org.dhis2.utils.customviews.navigationbar.NavigationPageConfigurator
import org.hisp.dhis.android.core.D2

@Module
class TeiDashboardModule(
    private val view: TeiDashboardContracts.View,
    val teiUid: String,
    val programUid: String?,
    private val enrollmentUid: String?,
    private val isPortrait: Boolean,
) {
    @Provides
    @PerActivity
    fun provideView(mobileActivity: TeiDashboardMobileActivity): TeiDashboardContracts.View = mobileActivity

    @Provides
    @PerActivity
    fun providePresenter(
        dashboardRepository: DashboardRepository,
        schedulerProvider: SchedulerProvider,
        analyticsHelper: AnalyticsHelper,
        preferenceProvider: PreferenceProvider,
        matomoAnalyticsController: MatomoAnalyticsController,
    ): TeiDashboardContracts.Presenter =
        TeiDashboardPresenter(
            view,
            programUid,
            dashboardRepository,
            schedulerProvider,
            analyticsHelper,
            preferenceProvider,
            matomoAnalyticsController,
        )

    @Provides
    @PerActivity
    fun provideEnrollmentConfiguration(
        d2: D2,
        dispatcher: DispatcherProvider,
    ) = enrollmentUid?.let { EnrollmentConfiguration(d2, it, dispatcher) }

    @Provides
    @PerActivity
    fun provideDateEditionWarningHandler(
        enrollmentConfiguration: EnrollmentConfiguration?,
        eventResourcesProvider: EventResourcesProvider,
    ) = DateEditionWarningHandler(
        enrollmentConfiguration,
        eventResourcesProvider,
    )

    @Provides
    @PerActivity
    fun dashboardRepository(
        d2: D2,
        charts: Charts,
        preferenceProvider: PreferenceProvider,
        teiAttributesProvider: TeiAttributesProvider,
        metadataIconProvider: MetadataIconProvider,
        programConfigurationRepository: ProgramConfigurationRepository,
        featureConfigRepository: FeatureConfigRepository,
    ): DashboardRepository =
        DashboardRepositoryImpl(
            d2,
            charts,
            teiUid,
            programUid,
            enrollmentUid,
            teiAttributesProvider,
            preferenceProvider,
            metadataIconProvider,
            programConfigurationRepository,
            featureConfigRepository,
        )

    @Provides
    @PerActivity
    fun ruleEngineRepository(d2: D2): RuleEngineHelper? {
        if (enrollmentUid.isNullOrEmpty()) return null
        return RuleEngineHelper(
            EvaluationType.Enrollment(enrollmentUid),
            org.dhis2.mobileProgramRules.RulesRepository(d2),
        )
    }

    @Provides
    @PerActivity
    fun pageConfigurator(dashboardRepository: DashboardRepository): NavigationPageConfigurator =
        TeiDashboardPageConfigurator(dashboardRepository, isPortrait)

    @Provides
    @PerActivity
    fun teiAttributesProvider(d2: D2): TeiAttributesProvider = TeiAttributesProvider(d2)

    @Provides
    @PerActivity
    fun providesViewModelFactory(
        repository: DashboardRepository,
        analyticsHelper: AnalyticsHelper,
        dispatcher: DispatcherProvider,
        pageConfigurator: NavigationPageConfigurator,
        resourcesManager: ResourceManager,
    ): DashboardViewModelFactory =
        DashboardViewModelFactory(
            repository,
            analyticsHelper,
            dispatcher,
            pageConfigurator,
            resourcesManager,
        )

    @Provides
    fun provideProgramConfigurationRepository(d2: D2): ProgramConfigurationRepository = ProgramConfigurationRepository(d2)
}
